File: services/spring/order-service/src/main/java/com/smartfulfillment/order_service/util/LocationUtils.java
package com.smartfulfillment.order_service.util;

public class LocationUtils {

    // Earth's mean radius in kilometers
    private static final int EARTH_RADIUS_KM = 6371;

    /**
     * Calculates the great-circle distance between two points on the Earth's surface.
     * Uses the Haversine formula:
     * a = sin²(Δφ/2) + cos φ1 ⋅ cos φ2 ⋅ sin²(Δλ/2)
     * c = 2 ⋅ atan2( √a, √(1−a) )
     * d = R ⋅ c
     */
    public static double calculateDistance(double startLat, double startLng, double endLat, double endLng){
        double dLat = Math.toRadians((endLat - startLat));
        double dLong = Math.toRadians((endLng - startLng));

        startLat = Math.toRadians(startLat);
        endLat = Math.toRadians(endLat);

        double a = Math.pow(Math.sin(dLat / 2), 2) +
                Math.cos(startLat) * Math.cos(endLat) *
                        Math.pow(Math.sin(dLong / 2), 2);

        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

        return EARTH_RADIUS_KM * c; // Distance in kilometers
    }
}

File: services/go/order-twin/internal/utils/location_utils.go
package utils

import "math"

const EarthRadiusKm = 6371.0

func CalculateDistance(startLat, startLng, endLat, endLng float64) float64 {
	dLat := (endLat - startLat) * math.Pi / 180.0
	dLng := (endLng - startLng) * math.Pi / 180.0

	startLatRad := startLat * math.Pi / 180.0
	endLatRad := endLat * math.Pi / 180.0

	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(startLatRad)*math.Cos(endLatRad)*math.Sin(dLng/2)*math.Sin(dLng/2)

	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))

	return EarthRadiusKm * c
}
